libraries/botframework-schema/src/index.ts

```diff
@@ -17,939 +19,1122 @@ export * from './teams';
+    speak?: string;
+    /**
+     * Indicates whether your bot is accepting,
+     * expecting, or ignoring user input after the message is delivered to the client. Possible
+     * values include: 'acceptingInput', 'ignoringInput', 'expectingInput'
+     */
+    inputHint?: InputHints | string;
+    /**
+     * The text to display if the channel cannot render cards.
+     */
+    summary?: string;
+    /**
+     * The suggested actions for the activity.
+     */
+    suggestedActions?: SuggestedActions;
+    /**
+     * Attachments
+     */
+    attachments?: Attachment[];
+    /**
+     * Represents the entities that were mentioned in the message.
+     */
+    entities?: Entity[];
+    /**
+     * Contains channel-specific content.
+     */
+    channelData?: any;
+    /**
+     * Indicates whether the recipient of a contactRelationUpdate was added or removed from the
+     * sender's contact list.
+     */
+    action?: string;
+    /**
+     * Contains the ID of the message to which this message is a reply.
+     */
+    replyToId?: string;
+    /**
+     * A descriptive label for the activity.
+     */
+    label: string;
+    /**
+     * The type of the activity's value object.
+     */
+    valueType: string;
+    /**
+     * A value that is associated with the activity.
+     */
+    value?: any;
+    /**
+     * The name of the operation associated with an invoke or event activity.
+     */
+    name?: ActivityEventNames | string;
+    /**
+     * A reference to another conversation or activity.
+     */
+    relatesTo?: ConversationReference;
+    /**
+     * The a code for endOfConversation activities that indicates why the conversation ended.
+     * Possible values include: 'unknown', 'completedSuccessfully', 'userCancelled', 'botTimedOut',
+     * 'botIssuedInvalidMessage', 'channelFailed'
+     */
+    code?: EndOfConversationCodes | string;
+    /**
+     * The time at which the activity should be considered to be "expired" and should not be
+     * presented to the recipient.
+     */
+    expiration?: Date;
+    /**
+     * The importance of the activity. Possible values include: 'low', 'normal', 'high'
+     */
+    importance?: ActivityImportance | string;
+    /**
+     * A delivery hint to signal to the recipient alternate delivery paths for the activity.
+     * The default delivery mode is "default". Possible values include: 'normal', 'notification', 'expectReplies', 'ephemeral'
+     */
+    deliveryMode?: DeliveryModes | string;
+    /**
+     * List of phrases and references that speech and language priming systems should listen for
+     */
+    listenFor: string[];
+    /**
+     * The collection of text fragments to highlight when the activity contains a ReplyToId value.
+     */
+    textHighlights?: TextHighlight[];
+    /**
+     * An optional programmatic action accompanying this request
+     */
+    semanticAction?: SemanticAction;
+}
+
+export const assertActivity: Assertion<Activity> = (value, path) => {
```

Comment:
I like this method, but am worried about the enforceability for it and repercussions for botbuilder-js dependent libraries that do not directly/robustly follow the Bot Framework protocol. E.g. what if some other library sends a non-normative `Activity.conversation` value. 

We might have to go back the drawing board and start with steadfastly following the [Basic Activity structure](https://github.com/microsoft/botframework-sdk/blob/main/specs/botframework-activity/botframework-activity.md#basic-activity-structure) defined in the spec.

Reply:
In cases where "bad" inputs are permissive, we need to ensure the type annotations are correct. Anywhere that doesn't require a full `Activity` should accept a `Partial<Activity>`, and fields can be (conditionally) extracted and asserted as necessary.

We may have been a little less strict in the past, but ultimately this leads to a bad experience for customers because we will assume we have type safety when we don't and introduce bugs as a result.

In this particular case, we _need_ a complete assertion for the full type in order to make the partial assertion. This partial assertion can then be used when we don't expect a fully constructed `Activity`.

Reply:
That sounds more than reasonable, what will happen is that we'll need to be more specific about when `"assertPartial*"` and `assert*` are used. We might have some growing pains in the 4.12 release and after,  but I think stricter is the way to go.

Reply:
Agreed - there will definitely be pain during the transition. My thinking is that we will likely move more towards using `Partial<Activity>` as a function parameter type and then doing something like:

```typescript
function activityThatShouldHaveTimestamp(activity: Partial<Activity>); //declare type as partial activity
function activityThatShouldHaveTimestamp(activity: unknown) { // assertions to narrow it down
    assertPartialActivity(activity);
    assert.date(activity.timestamp, ['activity', 'timestamp']);
    // activity.timestamp is now definitely an instance of a Date
}
```

This would yield a `TypeError` for an activity without a timestamp. It also doesn't rely on creating a (strictly complete) activity. It would work if the user passed in `{ timestamp: new Date() }`.

Reply:
Sounds good to me!

@benbrown @Batta32 @iMicknl FYI for [Botkit](https://github.com/howdyai/botkit), [microsoft/botframework-solutions](https://github.com/microsoft/botframework-solutions), [botbuilder-community](https://github.com/BotBuilderCommunity/), respectively